flocker/control/test/test_persistence.py

```diff
@@ -249,3 +333,202 @@ def test_complex_keys(self):
+    app_manifestations = set(app.volume.manifestation for app in applications)
+    # A set that contains all of those, plus an arbitrary set of
+    # manifestations.
+    dataset_ids = frozenset(
+        app.volume.manifestation.dataset_id
+        for app in applications if app.volume
+    )
+    manifestations = (
+        st.sets(MANIFESTATIONS.filter(
+            lambda m: m.dataset_id not in dataset_ids))
+        .map(pset)
+        .map(lambda ms: ms.union(app_manifestations))
+        .map(lambda ms: dict((m.dataset.dataset_id, m) for m in ms)))
+    return st.builds(
+        Node, uuid=UUIDS,
+        applications=st.just(applications),
+        manifestations=manifestations)
+
+
+NODES = st.sets(APPLICATIONS_WITH_VOLUMES).map(
+    lambda apps: pset(dict((
+        app.volume.manifestation.dataset_id, app) for app in apps).values())
+).flatmap(_build_node)
+
+
+DEPLOYMENTS = st.builds(
+    Deployment, nodes=st.sets(NODES, min_size=0, max_size=3)
+)
+
+
+class ConfigurationMigrationTests(SynchronousTestCase):
+    """
+    Tests for ``ConfigurationMigration`` class that performs individual
+    configuration upgrades.
+
+    There should be
+    """
+    @given(DEPLOYMENTS)
+    def test_upgrade_configuration_v1_latest(self, deployment):
+        """
+        Test a range of generated configurations (deployments) can be
+        upgraded from v1 to the latest configuration version.
+
+        This test will need updating for each new configuration version
+        introduced to reflect the expected configuration format
+        (``expected_configuration``) after a successful upgrade.
+        """
+        source_json = wire_encode(deployment)
+        upgraded_json = migrate_configuration(
+            1, _CONFIG_VERSION, source_json, ConfigurationMigration)
+        upgraded_config = wire_decode(upgraded_json)
+        expected_configuration = Configuration(
+            version=_CONFIG_VERSION, deployment=deployment
+        )
+        self.assertEqual(upgraded_config, expected_configuration)
+
+    @given(st.tuples(
+        st.integers(min_value=1, max_value=2),
+        st.integers(min_value=2, max_value=2)
+    ))
+    def test_upgrade_configuration_versions(self, versions):
+        """
+        Test a range of version upgrades by ensuring the configuration
+        blob after upgrade matches that which is expected for the
+        particular version.
+
+        See flocker/control/test/configurations for individual
+        version JSON files and generation code.
+
+        Increment the given integer values in the decorator above to define
+        the test range when new configuration upgraders are added. The first
+        range is the "upgrade from" versions range, where min_value should
+        always be 1. The second range is the "upgrade to" versions range,
+        where min_value should be 2 and max_value should be the latest
+        supported configuration version number.
```

Comment:
I very much appreciate that you've explained where the numbers come from. Thank you.

If I understand correctly, we already have a variable in code that says what the latest supported config version is: `_CONFIG_VERSION`. If that's right, we should use that, rather than a literal 2.

Second, I don't think we support downgrades. This test only passes because we have two versions. The real constraint is **not** that the "upgrade to" value should be more than 2, it's that it should be more than "upgrade from".

As such, a better approach would be to do something like this:

```python
SUPPORTED_VERSIONS = integers(1, _CONFIG_VERSION)

@given(tuples(SUPPORTED_VERSIONS, SUPPORTED_VERSIONS).map(lambda x: tuple(sorted(x)))
def test_u_c_v(self, (upgrade_from, upgrade_to)):
    ....
```